packages/contracts/src/core/dao/DAO.sol

```diff
@@ -182,23 +185,41 @@ contract DAO is
         execResults = new bytes[](_actions.length);
 
         for (uint256 i = 0; i < _actions.length; ) {
-            address to = _actions[i].to;
-            (bool success, bytes memory response) = to.call{value: _actions[i].value}(
-                _actions[i].data
-            );
-
-            if (!success) {
-                // If the call failed and wasn't allowed in allowFailureMap, revert.
-                if (!hasBit(_allowFailureMap, uint8(i))) {
+            // Check if failure is allowed
+            if (!hasBit(_allowFailureMap, uint8(i))) {
+                (bool success, bytes memory result) = _actions[i].to.call{value: _actions[i].value}(
+                    _actions[i].data
+                );
+
+                // Check if the call failed.
+                if (!success) {
                     revert ActionFailed(i);
                 }
 
-                // If the call failed, but was allowed in allowFailureMap, store that
-                // this specific action has actually failed.
-                failureMap = flipBit(failureMap, uint8(i));
-            }
+                execResults[i] = result;
+            } else {
+                uint256 gasBefore = gasleft(); // gasleft: 2 gas, assignment:
+
+                (bool success, bytes memory result) = _actions[i].to.call{value: _actions[i].value}(
+                    _actions[i].data
+                );
+                uint256 gasAfter = gasleft();
+
+                // Check if the call failed.
+                if (!success) {
+                    // Make sure that the action call did not fail because 63/64 of `gasleft()` was insufficient to execute the external call `.to.call` (see https://eips.ethereum.org/EIPS/eip-150).
+                    // In specific scenarios, i.e. proposal execution where the last action in the action array is allowed to fail, the account calling `execute` could force-fail this action by setting a gas limit
+                    // where 63/64 is insufficient causing the `.to.call` to fail, but where the remaining 1/64 gas are sufficient to successfully finish the `execute` call.
+                    if (gasAfter < gasBefore / 64) {
+                        revert InsufficientGas();
+                    }
+
+                    // Store that this action failed.
+                    failureMap = flipBit(failureMap, uint8(i));
+                }
 
-            execResults[i] = response;
+                execResults[i] = result;
```

Comment:
After checking how much gas `gasleft()` uses I suggest you switch it to always store `uint256 gasBefore = gasleft();` because this line uses ~5 gas.

Then you can write the code cleaner and safe us some gas during deploy time.

Reply:
Changed in f64efe79f3e40fee8e1227c32b2707e0568eca2e